src/docfx/restore/RestoreGit.cs

```diff
@@ -144,20 +148,40 @@ async Task AddWorkTrees()
                         var gitDependencyLock = dependencyLock?.GetGitLock(remote, branch);
                         headCommit = gitDependencyLock?.Commit ?? headCommit;
 
-                        var workTreeHead = $"{GetWorkTreeHeadPrefix(branch, !string.IsNullOrEmpty(gitDependencyLock?.Commit))}{headCommit}";
-                        var workTreePath = Path.GetFullPath(Path.Combine(repoPath, "../", workTreeHead)).Replace('\\', '/');
+                        var (workTreePath, gitSlot) = await DependencyGitPool.AcquireExclusiveGit(remote, branch, headCommit);
+                        workTreePath = Path.GetFullPath(workTreePath).Replace('\\', '/');
+                        var restored = true;
 
-                        if (existingWorkTreePath.TryAdd(workTreePath))
+                        try
                         {
-                            try
+                            if (existingWorkTreePath.TryAdd(workTreePath))
                             {
-                                await GitUtility.AddWorkTree(repoPath, headCommit, workTreePath);
+                                // create new worktree
+                                try
+                                {
+                                    await GitUtility.AddWorkTree(repoPath, headCommit, workTreePath);
+                                }
+                                catch (Exception ex)
+                                {
+                                    throw Errors.GitCloneFailed(remote, branches).ToException(ex);
+                                }
                             }
-                            catch (Exception ex)
+                            else
                             {
-                                throw Errors.GitCloneFailed(remote, branches).ToException(ex);
+                                // worktree already exists
+                                // checkout to {headCommit}, no need to fetch
+                                await GitUtility.Checkout(workTreePath, headCommit);
```

Comment:
Is add worktree enough to update `workTreePath` to `headCommit`? since we are calling `worktree add --force`

Reply:
no, it will throw work tree already exists error, because the work tree name is not changed